service/src/main/java/bio/terra/catalog/service/DatasetService.java

```diff
@@ -115,49 +116,37 @@ private ObjectNode toJsonNode(String json) {
     }
   }
 
-  private List<DatasetResponse> createDatasetResponses(
-      Map<String, StorageSystemInformation> datasetStorageInformation,
-      StorageSystem storageSystem) {
-    List<Dataset> datasets = datasetDao.find(storageSystem, datasetStorageInformation.keySet());
-    return datasets.stream()
-        .map(
-            dataset ->
-                new DatasetResponse(
-                    dataset, datasetStorageInformation.get(dataset.storageSourceId())))
-        .toList();
-  }
-
-  private List<DatasetResponse> collectDatasets(StorageSystem system) {
-    // For this storage system, get the collection of visible datasets and the user's roles for
-    // each dataset.
-    Map<String, StorageSystemInformation> roleMap = getService(system).getDatasets();
-    return createDatasetResponses(roleMap, system);
-  }
-
   public DatasetsListResponse listDatasets() {
-    List<DatasetResponse> datasets =
+    var systemsAndInfo =
         RequestContextCopier.parallelWithRequest(Arrays.stream(StorageSystem.values()))
-            .flatMap(system -> collectDatasets(system).stream())
-            .collect(Collectors.toList());
+            .collect(
+                Collectors.toMap(Function.identity(), system -> getService(system).getDatasets()));
 
+    List<Dataset> datasets;
     if (samService.hasGlobalAction(SamAction.READ_ANY_METADATA)) {
-      datasets.addAll(
-          datasetDao.listAllDatasets().stream()
-              .map(
-                  dataset ->
-                      new DatasetResponse(
-                          dataset, new StorageSystemInformation(DatasetAccessLevel.READER)))
-              .filter(
-                  datasetWithAccessLevel ->
-                      !datasets.stream()
-                          .map(datasetInList -> datasetInList.dataset.id())
-                          .toList()
-                          .contains(datasetWithAccessLevel.dataset.id()))
-              .toList());
+      datasets = datasetDao.listAllDatasets();
+    } else {
+      datasets =
+          datasetDao.find(
+              systemsAndInfo.entrySet().stream()
+                  .collect(
+                      Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().keySet())));
     }
     var response = new DatasetsListResponse();
-
-    response.setResult(datasets.stream().map(DatasetResponse::convertToObject).toList());
+    // This is used for the admin user, to provide a dummy role for datasets that an admin
+    // doesn't have access to in the underlying storage system.
+    var defaultInformation = new StorageSystemInformation(DatasetAccessLevel.READER);
+    response.setResult(
+        datasets.stream()
+            .map(
+                dataset ->
+                    new DatasetResponse(
+                        dataset,
+                        systemsAndInfo
+                            .getOrDefault(dataset.storageSystem(), Map.of())
```

Comment:
TOL: I wonder if it would be worth adding an 'UNKNOWN' storageSystem. Not for this PR, but maybe a ticket if we think it is a good idea.

Reply:
I'd be interested to hear more about how we would use it. I think unknown objects are good for cases where you want to avoid `null` checks and to provide default behavior. But since `getOrDefault()` is needed to handle the admin user case I'm not sure how an unknown storage system would be used here.

Reply:
Yeah! Of course! I was thinking it might make it easier for clients to use accessLevel if they can assume it is a non-nullable field (even for admins). That seems like a very small use though, and not necessarily worth the time to add.